### PR TITLE
adding ApplicationLoadBalancer module into terraform

### PR DESCRIPTION
**Pull request by prem-fissionhq**
https://dimagi-dev.atlassian.net/browse/SAAS-12150
##### ENVIRONMENTS AFFECTED

Staging

Launched proxycouch ALB and attached couch6 as the target on port: 15984. 

Please note: Inventory is not updated yet.

### CODE REVIEW

File: src/commcare_cloud/commands/terraform/templates/commcarehq.tf.j2 [hunk 1]
```diff
@@ -289,6 +289,27 @@ module "pgbouncer_nlb__{{ pgbouncer_nlb.name }}" {
 
 {% endfor %}
 
+{% for proxy_alb in proxy_albs %}
+module "proxy_alb__{{ proxy_alb.name }}" {
+  source = "./modules/proxy_alb"
```
Comment (dannyroberts) [inline src/commcare_cloud/commands/terraform/templates/commcarehq.tf.j2:294]: Like my comment above, I think we want to call the module either `couch_alb` or `internal_alb` depending on the direction selected above

File: environments/staging/terraform.yml [hunk 2]
```diff
@@ -194,11 +194,9 @@ servers:
     network_tier: "db-private"
     az: "b"
     volume_size: 30
-    volume_type: "gp3"
     volume_encrypted: yes
     block_device:
       volume_size: 80
-      volume_type: "gp3"
```
Comment (dannyroberts) [inline environments/staging/terraform.yml:201]: I think this is an unrelated change. Maybe put this in a different PR if you still want to make it?

File: environments/staging/terraform.yml [hunk 3]
```diff
@@ -395,6 +393,11 @@ pgbouncer_nlbs:
     targets:
       - pgbouncer0-staging
 
+proxy_albs:
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply environments/staging/terraform.yml:396]: Agree with you Danny. Yes, if we can have this as `internal_albs` it will be a good idea. I will be working on including them. 
Thank you.

File: environments/staging/terraform.yml [hunk 4]
```diff
@@ -395,6 +393,11 @@ pgbouncer_nlbs:
...
+  - name: "proxycouch_alb-staging"
```
[the remark this answers is not in the document]
Comment (prem-fissionhq) [reply environments/staging/terraform.yml:397]: Sure Danny.

File: environments/staging/terraform.yml (hunk 2 above)
Comment (prem-fissionhq) [reply environments/staging/terraform.yml:201]: yes, Danny. Sorry for pushing this change. Initially, it should lot of files are changes during that time I have missed this. Will fix this in the next commits.

File: environments/staging/aws-resources.yml [hunk 5]
```diff
@@ -8,6 +8,11 @@ couch15-staging: 10.201.42.109
 couch15-staging.instance_id: i-0f618ee36887bb494
 couch6-staging: 10.201.40.74
 couch6-staging.instance_id: i-01b32f8a2eebe07bc
+couch7-staging: 10.201.40.221
+couch7-staging.instance_id: i-0361f071191cd66c0
+couch8-staging: 10.201.40.18
+couch8-staging.instance_id: i-06d9f01ac9c8cb6b8
```
Comment (dannyroberts) [inline environments/staging/aws-resources.yml:14]: This looks like an unrelated change (I think these are the marches that you already terminated?)

File: environments/staging/terraform.yml [hunk 6]
```diff
@@ -157,6 +157,30 @@ servers:
...
+    server_instance_type: "t3.large"
+    network_tier: "db-private"
+    az: "b"
+    volume_size: 30
+    volume_encrypted: yes
+    block_device:
+      volume_size: 80
+      encrypted: true
+    group: "couchdb2"
+    os: ubuntu_pro_bionic
```
Comment (dannyroberts) [inline environments/staging/terraform.yml:182]: Was it intentional to add these two back?

File: src/commcare_cloud/ansible/group_vars/all.yml [hunk 7]
```diff
@@ -45,10 +45,11 @@ zookeeper_client_port: 2181
 formplayer_port: 8181
 
 couchdb_use_haproxy: False
+couchdb_use_alb: True
 couchdb_version: 2.3.1
 couchdb2_port: 15984
 couchdb2_local_port: "{{ 15986 if couchdb_version is version('3.0.0', '<') else couchdb2_port }}"
-couchdb2_proxy_port: "{{ 35984 if couchdb_use_haproxy else 25984 }}"
+couchdb2_proxy_port: "{{ 35984 if couchdb_use_haproxy else 80 if couchdb_use_alb else 25984 }}"
```
Comment (dannyroberts) [inline src/commcare_cloud/ansible/group_vars/all.yml:52]: Can you make the ALB listen on port 25984 or just the standard 5984? I think that would be a cleaner change

File: src/commcare_cloud/ansible/group_vars/all.yml [hunk 8]
```diff
@@ -45,10 +45,11 @@ zookeeper_client_port: 2181
 formplayer_port: 8181
 
 couchdb_use_haproxy: False
+couchdb_use_alb: True
```
Comment (dannyroberts) [inline src/commcare_cloud/ansible/group_vars/all.yml:48]: I think this should default to false. Defaults should be set to what third parties are most likely using / changes should be done so as not to affect them.

File: src/commcare_cloud/terraform/modules/internal_alb/main.tf [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+resource "aws_lb_target_group_attachment" "this" {
+  count = "${length(var.server_ids)}"
+  target_group_arn = "${aws_lb_target_group.this.arn}"
+  target_id        = "${var.server_ids[count.index]}"
+  port             = "${var.port}"
+}
+
+resource "aws_lb_listener" "this" {
+  load_balancer_arn = "${aws_lb.this.arn}"
+  port              = "80"
```
Comment (dannyroberts) [inline src/commcare_cloud/terraform/modules/internal_alb/main.tf:45]: I think this is where we want to change the port. We can either have it be whatever is set for `${var.port}` or have a separate variable for what port the alb should listen at.

**Review by dannyroberts — COMMENTED** [review]
This is looking really good. I had a few comments that I think would get this PR ready for prime time

File: src/commcare_cloud/ansible/group_vars/all.yml (hunk 8 above)
Comment (dannyroberts) [reply src/commcare_cloud/ansible/group_vars/all.yml:48]: (it can then be set to true in staging/public.yml)

**Comment by prem-fissionhq** [comment]
@dannyroberts thanks. I will work on the suggested changes.

File: src/commcare_cloud/ansible/group_vars/all.yml (hunk 8 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/ansible/group_vars/all.yml:48]:
```
Previous: couchdb2_proxy_port: "{{ 35984 if couchdb_use_haproxy else 25984 }}"
Current Change: couchdb2_proxy_port: "{{ 35984 if couchdb_use_haproxy else 80 if couchdb_use_alb else 25984 }}
```
 Danny: Can you make the ALB listen on port 25984 or just the standard 5984? I think that would be a cleaner change
Prem: Yes, we can do that then I think having an additional variable `couchdb_use_alb` is not required, Can I remove it?

File: src/commcare_cloud/terraform/modules/internal_alb/main.tf (hunk 9 above)
Comment (prem-fissionhq) [reply src/commcare_cloud/terraform/modules/internal_alb/main.tf]: Done.

**Comment by prem-fissionhq** [comment]
Created a new PR for this Change. Closing this PR.
PR: https://github.com/dimagi/commcare-cloud/pull/4764
